### PR TITLE
fix random OSD daemon pod failures when running OSD on PVC

### LINKED ISSUE #3612: [OSD on PVC] OSD fails randomly while running on PVC
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
- While running multiple OSD on top of PVCs, some of the OSD fail randomly. 

**Expected behavior:**
- All the OSD should run successfully without failure.

**How to reproduce it (minimal and precise):**
1. Create the cluster environment. (AWS or any local setup with Vagrant). 
2. Have Persistence Volume and Persistence Volume Claims based on the number of OSDs to be created. 
3. Apply the common.yaml file. 
4. Apply the operator.yaml file. 
5. Apply the cluster-on-pvc.yaml file with 'count', say, 10 (to start 10 OSD)
6. Observe the OSD prepare PODs. 
- Some (1 or 2) of the OSD prepare PODs will restart and ultimately fail and corresponding OSDs will never get created

Error logs:

- OSD Prepare POD logs:

```
2019-08-07 10:01:43.930960 I | cephcmd: desired devices to configure osds: [{Name:/mnt/cluster1-set1-3-0-template-1-pp2t2 OSDsPerDevice:1 IsFilter:false}]
2019-08-07 10:01:43.934091 I | rookcmd: starting Rook v1.0.0-318.g031c3bd7.dirty with arguments '/rook/rook ceph osd provision'
2019-08-07 10:01:43.934109 I | rookcmd: flag values: --cluster-id=e019bc7a-4783-40cd-a253-bf7e0c4004f4, --data-device-filter=, --data-devices=/mnt/cluster1-set1-3-0-template-1-pp2t2, --data-directories=, --encrypted-device=false, --force-format=false, --help=false, --location=, --log-flush-frequency=5s, --log-level=INFO, --metadata-device=, --node-name=cluster1-set1-3-0-template-1-pp2t2, --operator-image=, --osd-database-size=20480, --osd-journal-size=5120, --osd-store=, --osd-wal-size=576, --osds-per-device=1, --pvc-backed-osd=true, --service-account=
2019-08-07 10:01:43.934116 I | op-mon: parsing mon endpoints: c=10.109.221.193:6789,a=10.96.84.1:6789,b=10.110.117.47:6789
2019-08-07 10:01:43.958412 W | cephconfig: failed to add config file override from '/etc/rook/config/override.conf': open /etc/rook/config/override.conf: no such file or directory
2019-08-07 10:01:43.958427 I | cephconfig: writing config file /var/lib/rook/rook-ceph/rook-ceph.config
2019-08-07 10:01:43.958527 I | cephconfig: copying config to /etc/ceph/ceph.conf
2019-08-07 10:01:43.958644 I | cephconfig: generated admin config in /var/lib/rook/rook-ceph
2019-08-07 10:01:43.958649 I | cephosd: discovering hardware
2019-08-07 10:01:43.958665 I | exec: Running command: lsblk /mnt/cluster1-set1-3-0-template-1-pp2t2 --bytes --nodeps --pairs --output SIZE,ROTA,RO,TYPE,PKNAME
2019-08-07 10:01:43.969180 I | exec: Running command: sgdisk --print /mnt/cluster1-set1-3-0-template-1-pp2t2
2019-08-07 10:01:43.972595 I | cephosd: creating and starting the osds
2019-08-07 10:01:43.972614 I | exec: Running command: lsblk /mnt/cluster1-set1-3-0-template-1-pp2t2 --bytes --pairs --output NAME,SIZE,TYPE,PKNAME
2019-08-07 10:01:43.975214 I | cephosd: /mnt/cluster1-set1-3-0-template-1-pp2t2 found in the desired devices
2019-08-07 10:01:43.984976 I | cephosd: configuring osd devices: {"Entries":{"/mnt/cluster1-set1-3-0-template-1-pp2t2":{"Data":-1,"Metadata":null,"Config":{"Name":"/mnt/cluster1-set1-3-0-template-1-pp2t2","OSDsPerDevice":1,"IsFilter":false},"LegacyPartitionsFound":false}}}
2019-08-07 10:01:43.993032 I | cephosd: device /mnt/cluster1-set1-3-0-template-1-pp2t2 to be configured by ceph-volume
2019-08-07 10:01:43.993058 I | cephosd: 0/0 pre-ceph-volume osd devices succeeded on this node
2019-08-07 10:01:43.993377 I | exec: Running command: ceph auth get-or-create-key client.bootstrap-osd mon allow profile bootstrap-osd --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json --out-file /tmp/168398674
2019-08-07 10:01:44.599838 I | exec: Running command: stdbuf -oL sed -i -e s#udev_sync = 1#udev_sync = 0# /etc/lvm/lvm.conf
2019-08-07 10:01:44.607965 I | exec: Running command: stdbuf -oL sed -i -e s#udev_rules = 1#udev_rules = 0# /etc/lvm/lvm.conf
2019-08-07 10:01:44.618415 I | exec: Running command: stdbuf -oL sed -i -e s#use_lvmetad = 1#use_lvmetad = 0# /etc/lvm/lvm.conf
2019-08-07 10:01:44.627997 I | exec: Running command: stdbuf -oL sed -i -e s#scan = \[ "/dev" \]#scan = [ "/dev", "/mnt" ]# /etc/lvm/lvm.conf
2019-08-07 10:01:44.635056 I | exec: Running command: stdbuf -oL sed -i -e 0,/# filter =.*/{s%# filter =.*% filter = [ "a|^/mnt/.*| r|.*/|" ]%} /etc/lvm/lvm.conf
2019-08-07 10:01:44.646724 I | cephosd: configuring new device /mnt/cluster1-set1-3-0-template-1-pp2t2
2019-08-07 10:01:44.646744 I | exec: Running command: stdbuf -oL ceph-volume lvm prepare --no-systemd --data /mnt/cluster1-set1-3-0-template-1-pp2t2
2019-08-07 10:01:59.552106 I | cephosd: SP - failed ceph-volume. Failed to complete '': exit status 1. -->  RuntimeError: Command failed with exit code 250: /bin/ceph-osd --cluster ceph --osd-objectstore bluestore --mkfs -i 3 --monmap /var/lib/ceph/osd/ceph-3/activate.monmap --keyfile - --osd-data /var/lib/ceph/osd/ceph-3/ --osd-uuid 9e94854a-9662-4041-b0e2-229ae8a2940b --setuser ceph --setgroup ceph
```
**File(s) to submit**:
ceph-volume-logs (2).txt

**Environment**:
* OS (e.g. from /etc/os-release):
* Kernel (e.g. `uname -a`):
* Cloud provider or hardware configuration:
* Rook version (use `rook version` inside of a Rook Pod):
* Storage backend version (e.g. for ceph do `ceph -v`):
* Kubernetes version (use `kubectl version`): v1.15.1
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)):

### PR DESCRIPTION
**Pull request by sp98**
Signed-off-by: Santosh Pillai <sapillai@redhat.com>

**Description of your changes:**
 OSD pods when running on PVC fail randomly due to some race condition:
 OSD daemon pod was activating the volume group using vgchange -an without any volume group name.
 - This caused the pods to fail randomly when number of OSDs is more, say greater than 10.
 - This fix adds Volume group name while activating/deactivating volume groups to fix this

**Which issue is resolved by this Pull Request:**
Resolves #3612

**Checklist:**

- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [ ] Add the flag for skipping the CI if this PR does not require a build. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for more details.

### CODE REVIEW

**Comment by sp98** [comment]
I've tested this change by:
1. Creating 30 OSD on PVC
2. Updating OSD count from 15 to 25. 

`
csi-cephfsplugin-2wlb2                           3/3     Running     0          28m
csi-cephfsplugin-8n26h                           3/3     Running     0          28m
csi-cephfsplugin-b8gp8                           3/3     Running     0          28m
csi-cephfsplugin-provisioner-75f5955876-8vl4v    4/4     Running     0          28m
csi-cephfsplugin-provisioner-75f5955876-txsld    4/4     Running     0          28m
csi-rbdplugin-lmwd5                              3/3     Running     0          28m
csi-rbdplugin-nbs6m                              3/3     Running     0          28m
csi-rbdplugin-nj77k                              3/3     Running     0          28m
csi-rbdplugin-provisioner-76cf8977f8-p8fl4       5/5     Running     0          28m
csi-rbdplugin-provisioner-76cf8977f8-tfc5f       5/5     Running     0          28m
rook-ceph-agent-5ggn7                            1/1     Running     0          28m
rook-ceph-agent-9rpp7                            1/1     Running     0          28m
rook-ceph-agent-xws5d                            1/1     Running     0          28m
rook-ceph-mgr-a-868557bf56-k94l7                 1/1     Running     1          26m
rook-ceph-mon-a-676cf4c5cd-989mg                 1/1     Running     0          27m
rook-ceph-mon-b-54b474d67c-5smd2                 1/1     Running     0          27m
rook-ceph-mon-c-6c5845bc78-77dgw                 1/1     Running     0          26m
rook-ceph-operator-8484c8b749-wbq2l              1/1     Running     0          30m
rook-ceph-osd-0-5b8669476b-hsj5m                 1/1     Running     0          22m
rook-ceph-osd-1-6fc9b4579d-jsdkj                 1/1     Running     0          21m
rook-ceph-osd-10-57549fd5ff-2dv8x                1/1     Running     0          18m
rook-ceph-osd-11-66b65857c9-xmcvx                1/1     Running     0          18m
rook-ceph-osd-12-7769f77bb7-gjtv4                1/1     Running     0          18m
rook-ceph-osd-13-7f58c648f4-h2qhv                1/1     Running     0          18m
rook-ceph-osd-14-6c6cc68477-lgtzq                1/1     Running     0          18m
rook-ceph-osd-15-58db8b6449-sfkvn                1/1     Running     0          16m
rook-ceph-osd-16-796d49f77f-dht6r                1/1     Running     0          17m
rook-ceph-osd-17-5d9bf95695-zt7mz                1/1     Running     0          16m
rook-ceph-osd-18-d75dd98bb-db4ff                 1/1     Running     0          17m
rook-ceph-osd-19-79bff7f766-mkrb6                1/1     Running     0          16m
rook-ceph-osd-2-86bb6fb856-lg5nq                 1/1     Running     0          21m
rook-ceph-osd-20-59c65c458-dz8cj                 1/1     Running     0          15m
rook-ceph-osd-21-69f98c8fd5-pjgc2                1/1     Running     0          15m
rook-ceph-osd-22-f84556968-xgfrw                 1/1     Running     0          15m
rook-ceph-osd-23-84d667b67b-fqpz5                1/1     Running     0          15m
rook-ceph-osd-24-84c8ff68fc-d2pkl                1/1     Running     0          15m
rook-ceph-osd-25-6fb9cf54cb-qn87k                1/1     Running     0          15m
rook-ceph-osd-26-6979d879f5-dntp4                1/1     Running     0          14m
rook-ceph-osd-27-79ffbd7dd6-g67kj                1/1     Running     0          15m
rook-ceph-osd-28-649978c695-xvgxh                1/1     Running     0          14m
rook-ceph-osd-29-64b4bd6677-tb9gx                1/1     Running     0          14m
rook-ceph-osd-3-8498d74bc9-xknqq                 1/1     Running     0          19m
rook-ceph-osd-4-5d5c647bc6-rmcb4                 1/1     Running     0          19m
rook-ceph-osd-5-559845fc6f-5rqxw                 1/1     Running     0          21m
rook-ceph-osd-6-8576cd5bc4-hgjsb                 1/1     Running     0          21m
rook-ceph-osd-7-6b6789c65f-klrgp                 1/1     Running     0          21m
rook-ceph-osd-8-77fdb85b94-f25xs                 1/1     Running     0          21m
rook-ceph-osd-9-5758bf45fd-g85c6                 1/1     Running     0          18m
rook-ceph-osd-prepare-set1-0-data-qq764-t8vs6    0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-1-data-d5sxp-x89mb    0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-10-data-5ppzw-qcq6z   0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-11-data-c8fx4-8ngrk   0/1     OOMKilled   0          25m
rook-ceph-osd-prepare-set1-12-data-mtdb6-mkjdd   0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-13-data-qs7t8-qgdgv   0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-14-data-m9grr-2x886   0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-15-data-ccl2p-8zvgt   0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-16-data-bfxsw-gr2vj   0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-17-data-lgx2h-gls55   0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-18-data-qlpps-77q2h   0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-19-data-hrxts-blq2c   0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-2-data-mgth5-6kkvv    0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-20-data-qxx5c-zq285   0/1     OOMKilled   0          25m
rook-ceph-osd-prepare-set1-21-data-lvz6x-n4bxg   0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-22-data-zwxg4-l8ncd   0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-23-data-p2k9p-gc948   0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-24-data-8xp62-tk2xc   0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-25-data-2jw6j-4lk9r   0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-26-data-lvv7r-zncbp   0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-27-data-g9wn9-dj4q5   0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-28-data-f6ms9-z7p7l   0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-29-data-nchbp-5dgp2   0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-3-data-wq8pm-4ww44    0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-4-data-w9tdt-nl4tw    0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-5-data-qst7g-gqh8j    0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-6-data-lkjbl-f2lnl    0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-7-data-5rvb5-tbvgz    0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-8-data-9xwdq-5fzqt    0/1     Completed   0          25m
rook-ceph-osd-prepare-set1-9-data-kk5jt-jrnj5    0/1     OOMKilled   0          25m
rook-discover-5tbrb                              1/1     Running     0          29m
rook-discover-msnzl                              1/1     Running     0          29m
rook-discover-scqjk                              1/1     Running     0          29m
`

File: pkg/daemon/ceph/osd/daemon.go [hunk 1]
```diff
@@ -49,14 +49,25 @@ func StartOSD(context *clusterd.Context, osdType, osdID, osdUUID string, pvcBack
 		logger.Errorf("failed to create config dir %s. %+v", configDir, err)
 	}
 
+	var volumeGroupName string
 	if pvcBackedOSD {
+		logger.Info("PVC backed 2!!")
```
Comment (mykaul) [inline pkg/daemon/ceph/osd/daemon.go]: ?

File: pkg/daemon/ceph/osd/daemon.go [hunk 2]
```diff
@@ -49,14 +49,25 @@ func StartOSD(context *clusterd.Context, osdType, osdID, osdUUID string, pvcBack
...
+		volumeGroupName, err = oposd.GetVolumeGroupName(lvPath)
+		if err != nil {
+			return fmt.Errorf("error fetching volume group name for OSD %s. %+v", osdID, err)
+		}
+
 		if err := updateLVMConfig(context); err != nil {
```
Comment (mykaul) [inline pkg/daemon/ceph/osd/daemon.go]: Is the LVM config (on the host?) updated just because we manipulated some files? Don't you need to reload LVM or something? Perhaps use '--config' instead for the command?

File: pkg/daemon/ceph/osd/daemon.go (hunk 1 above)
Comment (sp98) [reply pkg/daemon/ceph/osd/daemon.go]: removed.

File: pkg/daemon/ceph/osd/daemon.go (hunk 2 above)
Comment (sp98) [reply pkg/daemon/ceph/osd/daemon.go]: -  lvm config is update on the osd-prepare and osd-daemon pod. Its not updated on the host. 
-  lvm changes that we did to make it work were mostly inspired by this [gist](https://gist.github.com/JohnStrunk/72953876611edea05be7fd302ea6e8ea)
- Need to check and get back to you regarding the latter part of the comment about reloading lv m and using `--config`

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 3]
```diff
@@ -785,3 +792,24 @@ func (c *Cluster) getOSDLabels(osdID int, failureDomainValue string, portable bo
 		portableKey:         strconv.FormatBool(portable),
 	}
 }
+
+//GetVolumeGroupName returns the Volume group name from the given Logical Volume Path
+func GetVolumeGroupName(lvPath string) (string, error) {
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/spec.go]: Please add a unit test for this method

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 4]
```diff
@@ -785,3 +792,24 @@ func (c *Cluster) getOSDLabels(osdID int, failureDomainValue string, portable bo
...
+	if lvPath == "" || strings.Split(lvPath, "/")[2] == "" {
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/spec.go]: After calling `strings.Split()`, need to check its length before calling `[2]`

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 5]
```diff
@@ -785,3 +792,24 @@ func (c *Cluster) getOSDLabels(osdID int, failureDomainValue string, portable bo
...
+
+//ReleaseLVMDevice deativates the LV to release the device.
+func ReleaseLVMDevice(context *clusterd.Context, volumeGroupName string) error {
+
+	if err := context.Executor.ExecuteCommand(false, "", "lvchange", "-an", volumeGroupName); err != nil {
+		return fmt.Errorf("failed to deactivate LVM %s. Error: %+v", volumeGroupName, err)
+	}
+
+	logger.Info("Successfully released device from lvm")
+
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/spec.go]: nit: extra line before returning looks better if removed

File: pkg/operator/ceph/spec/spec.go [hunk 6]
```diff
@@ -367,11 +367,13 @@ func generateLifeCycleCmd(dataDirHostPath string) []string {
 func PodLifeCycle(dataDirHostPath string) *v1.Lifecycle {
 	cmd := generateLifeCycleCmd(dataDirHostPath)
 
-	return &v1.Lifecycle{
+	lifeCyleHooks := &v1.Lifecycle{
```
Comment (travisn) [inline pkg/operator/ceph/spec/spec.go]: Was the lifecycle hook not needed since you added the release after the ceph-osd process exits? Is this change still needed?

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 7]
```diff
@@ -785,3 +792,24 @@ func (c *Cluster) getOSDLabels(osdID int, failureDomainValue string, portable bo
 		portableKey:         strconv.FormatBool(portable),
 	}
 }
+
+//GetVolumeGroupName returns the Volume group name from the given Logical Volume Path
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/spec.go]: Looks like this method is only called by the `daemon` package. How about if you move this method to that package?

Comment (sp98) [reply pkg/operator/ceph/cluster/osd/spec.go]: moved both `getVolumeGroupName` and `releaseLVMDevice` to `daemon` package as they are only used there

File: pkg/operator/ceph/spec/spec.go (hunk 6 above)
Comment (sp98) [reply pkg/operator/ceph/spec/spec.go]: Yes. `preStop` life cycle hook would be more relevant for fix in #3755. Reverted these changes back to original.

File: pkg/operator/ceph/cluster/osd/spec.go (hunk 3 above)
Comment (sp98) [reply pkg/operator/ceph/cluster/osd/spec.go]: added unit tests.

File: pkg/operator/ceph/cluster/osd/spec.go (hunk 4 above)
Comment (sp98) [reply pkg/operator/ceph/cluster/osd/spec.go]: updated. Adding more strict conditions for lv path format.
